radix-engine/src/utils/flash_state_generator.rs

```diff
@@ -224,3 +216,177 @@ pub fn generate_seconds_precision_state_updates<S: SubstateDatabase>(db: &S) ->
+                    ModuleId::Main,
+                    collection.collection_index(),
+                )
+                .unwrap();
+
+            state_updates
+                .by_node
+                .entry(node_id)
+                .or_default()
+                .of_partition(partition_num)
+                .delete()
+        }
+
+        // Create the new substates based on the new definition and code id.
+        let new_code_id = POOL_V1_1_CODE_ID;
+        let original_code = new_code_id.to_be_bytes().to_vec();
+
+        let package_structure = PackageNativePackage::validate_and_build_package_structure(
+            PoolNativePackage::definition(PoolV1MinorVersion::One),
+            VmType::Native,
+            original_code,
+            Default::default(),
+            &MockVmApi,
+        )
+        .unwrap();
+
+        let royalty_vault = reader
+            .read_object_field(
+                &node_id,
+                ModuleId::Main,
+                PackageField::RoyaltyAccumulator.field_index(),
+            )
+            .unwrap()
+            .as_typed::<PackageRoyaltyAccumulatorFieldPayload>()
+            .unwrap()
+            .into_latest()
+            .royalty_vault;
+
+        let node_substates = create_package_partition_substates(
+            package_structure,
+            metadata_init! {
+                "name" => "Pool Package".to_owned(), locked;
+                "description" => "A native package that defines the logic for a selection of pool components.".to_owned(), locked;
+            },
+            Some(royalty_vault),
+        );
+
+        node_substates
+            .into_iter()
+            .for_each(|(partition_number, entries)| {
+                state_updates
+                    .by_node
+                    .entry(node_id)
+                    .or_default()
+                    .of_partition(partition_number)
+                    .update_substates(
+                        entries
+                            .into_iter()
+                            .map(|(key, value)| (key, DatabaseUpdate::Set(value.into()))),
+                    );
+            })
+    }
+
+    /// Generates the state updates required for the new pool role.
+    ///
+    /// We have added a new role to the pool blueprints with v1.1 of the blueprints called the
+    /// `pool_contributor` role, which is the role that is allowed to contribute assets to the pool.
+    /// We would like pools that have already been instantiated to continue to function in the same
+    /// way as it did before the protocol update. As in, the protocol manager was the only role that
+    /// has contribution powers. So, for each pool that we find, we add a new role assignment for
+    /// the contributor role with the same rule as the pool manager. Thus, pools that were created
+    /// without this role will continue to function in the same way.
+    fn generate_role_assignment_update<S: SubstateDatabase + ListableSubstateDatabase>(
```

Comment:
woah, I did not realize we're doing whole scan + updates. Is there a better way to do this such as lazy updating? I'm concerned that this will take a bit of time to run if this is to be computed during the protocol update.